hazelcast/src/main/java/com/hazelcast/instance/impl/DefaultAddressPicker.java

```diff
@@ -381,30 +380,53 @@ AddressDefinition pickMatchingAddress(Collection<InterfaceDefinition> interfaces
                 if (address == null) {
                     continue;
                 }
-                matchingAddress = address;
-
-                if (preferIPv6Addresses) {
-                    // IPv6 address is preferred, return if address is IPv6.
-                    if (inetAddress instanceof Inet6Address) {
-                        return matchingAddress;
-                    }
-                } else if (inetAddress instanceof Inet4Address) {
-                    // No IPv6 address preference, return if address is IPv4.
-                    return matchingAddress;
+                if (getPriority(address, preferIPv6Addresses) > getPriority(matchingAddress, preferIPv6Addresses)) {
+                    matchingAddress = address;
                 }
             }
         }
         // nothing matched to IP version preference, return what we have.
         return matchingAddress;
     }
 
+    /**
+     * The priority in desc-order:
+     * <ol>
+     * <li>IPv6-preference
+     * <li>global address
+     * <li>site-local address
+     * <li>link-local address
+     * <li>loopback address
+     * </ol>
+     */
+    @SuppressWarnings("checkstyle:MagicNumber")
+    private int getPriority(AddressDefinition address, boolean preferIPv6Addresses) {
+        if (address == null) {
+            return 0;
+        }
+        InetAddress ia = address.inetAddress;
+        boolean isIpv6 = ia instanceof Inet6Address;
+        int priority;
+        if (ia.isLoopbackAddress()) {
+            priority = 1;
+        } else if (ia.isLinkLocalAddress()) {
+            priority = 2;
+        } else if (ia.isSiteLocalAddress()) {
+            priority = 3;
+        } else {
+            priority = 5;
+        }
+        if (isIpv6 ^ !preferIPv6Addresses) {
+            priority += 10;
+        }
```

Comment:
Looking back over this again, unless I'm missing something, I believe the XOR logic here is incorrect - the priority order explained in the Java-doc suggests that IPv6 addresses should have the highest priority (as the list is in descending order, with the highest priority listed first), and given the loopback address is given a priority of `1`, we can assume that a higher priority value means a higher priority for selection (also confirmed by usage in `pickMatchingAddress` function).

However, this XOR logic currently results in the priority being increased when `isIpv6 == true` and `preferIPv6Addresses == false`, which seems incorrect to me, as the priority is therefore **not increased** when the address is IPv6 and we prefer IPv6 addresses. Related tests may need to be adjusted as well if this logic is not correct.